test/opf/sdpwrm.jl

```diff
@@ -0,0 +1,222 @@
+    # TODO: use JuMP.primal_feasibility_report instead
+    #    (would require extracting a variable => value Dict)
+    sol_pm = res_pm["solution"]
+    var2val_pm = Dict(
+        :pg => Float64[sol_pm["gen"]["$g"]["pg"] for g in 1:G],
+        :qg => Float64[sol_pm["gen"]["$g"]["qg"] for g in 1:G],
+        :wm => Float64[sol_pm["bus"]["$i"]["w"] for i in 1:N]  # the diagonal elements of sol_pm["WR"]. Note that the rows and columns of sol_pm["WR"] are ordered in the order of ref[:bus].
+    )
+    model = opf.model
+    for varname in [:pg, :qg]
+        x = model[varname]
+        v = var2val_pm[varname]
+        @constraint(model, v .- 1e-3 .<= x .<= v .+ 1e-3)  # more tolerance for SDP
+    end
+    @constraint(model, var2val_pm[:wm] .- 1e-3 .<= diag(model[:WR]) .<= var2val_pm[:wm] .+ 1e-3)
+
+    optimize!(model)
+    @test termination_status(model) ∈ [OPTIMAL, ALMOST_OPTIMAL, LOCALLY_SOLVED, ALMOST_LOCALLY_SOLVED]
+    @test primal_status(model) ∈ [FEASIBLE_POINT, NEARLY_FEASIBLE_POINT]
+    # Also check that we get the same objective value as PowerModels
+    @test isapprox(objective_value(opf.model), res_pm["objective"], atol=1e-6, rtol=1e-6)
+
+    return nothing
+end
+
+"""
+    _test_sdpwrm_DualFeasibility()
+
+Test dual feasibility of SDPWRM problem.
+
+This test is executed on the 118 bus system.
+"""
+function _test_sdpwrm_DualFeasibility()
+    T = Float128
+    data = make_basic_network(pglib("pglib_opf_case118_ieee"))
+    solver = JuMP.optimizer_with_attributes(Clarabel.Optimizer{T},
+        "verbose" => true,
+        "equilibrate_enable" => false,
+        "tol_gap_abs"    => 1e-14,
+        "tol_gap_rel"    => 1e-14,
+        "tol_feas"       => 1e-14,
+        "tol_infeas_rel" => 1e-14,
+        "tol_ktratio"    => 1e-14,
+    )
+    opf = OPFGenerator.build_opf(SDPWRMPowerModel, data, solver; T=T)
+    # set_silent(opf.model)
+    OPFGenerator.solve!(opf)
+    res = OPFGenerator.extract_result(opf)
+
+    _test_sdpwrm_DualFeasibility(data, res)
+
+    return nothing
+end
+
+function _get_sym(i, j, v, N)
+    # return an N * N sparse symmetric matrix where the (i, j) entry is v
+    return Symmetric(sparse([i], [j], v, N, N))
+end
+
+function _get_skew_sym(i, j, v, N)
+    # return an N * N sparse skew-symmetric matrix where the (i, j) entry is v
+    return sparse([i, j], [j, i], [v, -v], N, N)
+end
+
+function _test_sdpwrm_DualFeasibility(data, res; atol=1e-6)
+    ref = PM.build_ref(data)[:it][:pm][:nw][0]
+    N = length(ref[:bus])
+    E = length(ref[:branch])
+    bus_loads = [
+        [ref[:load][l] for l in ref[:bus_loads][i]]
+        for i in 1:N
+    ]
+    # Bus-level data
+    bus_shunts = [
+        [ref[:shunt][s] for s in ref[:bus_shunts][i]]
+        for i in 1:N
+    ]
+    gs = [sum(shunt["gs"] for shunt in bus_shunts[i]; init=0.0) for i in 1:N]
+    bs = [sum(shunt["bs"] for shunt in bus_shunts[i]; init=0.0) for i in 1:N]
+
+    # Extract branch-level data
```

Comment:
That looks like boiler-plate code which could be useful to have in the source code